### ISSUE TITLE
[prefer-readonly-parameter-types] False positive with React.PropsWithChlidren

### ISSUE THREAD

**Issue by mdesantis** [typescript-eslint/typescript-eslint#4492]
I have something that I believe is a false positive with `React.PropsWithChildren`. The following function doesn't break the rule:

```tsx
function test(
  props: Readonly<Record<string, never>>
): JSX.Element {
  return <div/>
}
```

But the next one does:

```tsx
function test(
  props: Readonly<React.PropsWithChildren<Record<string, never>>>
): JSX.Element {
  return <div/>
}
```

And the next one too:

```tsx
function test(
  props: Readonly<React.PropsWithChildren<Readonly<Record<string, never>>>>
): JSX.Element {
    return <div/>
}
```

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

I have created a [reproduction sandbox on Glitch](https://glitch.com/edit/#!/dusty-six-fall): just open the terminal and run `npx eslint src/test.tsx`.

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-readonly-parameter-types": 2
  }
}
```

```tsx
function test(
  props: Readonly<React.PropsWithChildren<Record<string, never>>>
): JSX.Element {
  return <div/>
}
```

**Expected Result**

I expect that Eslint shouldn't report an error

**Actual Result**

Eslint reports an error:

```
$ npx eslint src/test.tsx 

/app/src/test.tsx
   8:3  error  Parameter should be a read only type  @typescript-eslint/prefer-readonly-parameter-types
  14:3  error  Parameter should be a read only type  @typescript-eslint/prefer-readonly-parameter-types

✖ 2 problems (2 errors, 0 warnings)

```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.10.1` |
| `@typescript-eslint/parser`        | `5.10.1` |
| `TypeScript`                       | `4.5.5` |
| `ESLint`                           | `8.8.0` |
| `node`                             | `16.12.0` |

**Comment by bradzacher** [comment]
Heads up that we have our own playground at https://typescript-eslint.io/

**Comment by mdesantis** [comment]
> Heads up that we have our own playground at https://typescript-eslint.io/

I noticed it, but:

1. I can't change package versions there
2. I can't reproduce the issue there

**Comment by bradzacher** [comment]
Looks like there is a bug in the playground which is breaking the rule - filed #4493.

As for glitch - I don't have the option to open the terminal, so I can't run your repro.
I copied the test code locally against master and it repros:
<img width="855" alt="image" src="https://user-images.githubusercontent.com/7462525/151678880-f594f72f-b001-4fea-a1ce-3fee99603ea7.png">

If I had to guess - it's likely one of the recent changes to the tooling that caused this regression.
cc @RebeccaStevens as I know your plugin depends on this tooling.

**Comment by bradzacher** [comment]
Actually looking deeper.... I think this is expected behaviour.
`PropsWithChildren` adds `children?: ReactNode` to your props.
`ReactNode` includes a number of *object* types which are not deeply readonly.
Thus your props are not deeply readonly - causing the rule to error.

It might seem counter-intuitive, but technically without making it deeply readonly - it's valid for you to do like `props.children[0].type = 'wrong'` within your code (if you narrow the types)
This rule does not enforce shallow readonly, it enforced deep readonly to ensure you cannot mutate the argument.

There's #4436 which adds an ignore list to the rule, which may be of interest if you want to ignore specific types and treat them as always readonly.

**Comment by mdesantis** [comment]
Thank you for the explanation, it feels so obvious now that you figured it out :D I'll try to use the type allowlist as you suggested. Good Saturday!
